runtime/include/cloe/data_broker.hpp

```diff
@@ -0,0 +1,1099 @@
+    assert(trigger_fn);
+    return *trigger_fn;
+  }
+
+ private:
+  /**
+   * Unpack the event to Event<const T*> and provide it to the caller.
+   *
+   * \tparam T Type of the event
+   * \param callback Caller function which accepts the unpacked event
+   * \note: In case type T and actual type do not match an exception is thrown
+   */
+  template <typename T>
+  void subscribe_impl(
+      const std::function<void(databroker::Event<databroker::signal_type_cref_t<T>>&)>& callback) {
+    typed_value_changed_event_t<T>* value_changed_event =
+        std::any_cast<typed_value_changed_event_t<T>>(&value_changed_event_);
+    if (callback) {
+      assert(value_changed_event);
+      callback(*value_changed_event);
+    }
+  }
+
+ public:
+  /**
+   * Subscribe to value-changed events.
+   *
+   * \tparam T Type of the signal
+   * \param callback event-function which will be called when the value changed
+   */
+  template <typename T>
+  void subscribe(databroker::on_value_changed_callback_t<T> callback) {
+    assert_static_type<T>();
+    assert_dynamic_type<T>();
+
+    subscribe_impl<T>(
+        [callback = std::move(callback)](typed_value_changed_event_t<T>& value_changed_event) {
+          value_changed_event.add(std::move(callback));
+        });
+  }
+
+  /**
+   * Return the count of subscribers to the value_changed event.
+   *
+   * \return size_t Count of subscribers to the value_changed event
+   */
+  std::size_t subscriber_count() const { return subscriber_count_(); }
+
+  /**
+   * Indicate whether the value_changed event has subscribers.
+   *
+   * \return bool True if the value_changed event has subscribers, false otherwise
+   */
+  bool has_subscriber() const { return subscriber_count() > 0; }
+
+  /**
+   * Return the list of names assigned to the signal.
+   *
+   * \return List of names assigned to the signal
+   */
+  const std::vector<std::string>& names() const { return names_; }
+
+  /**
+   * Return the first assigned name of the signal.
+   *
+   * \return First name of the signal
+   */
+  const std::string& name() const {
+    if (!names_.size()) {
+      throw std::logic_error("signal does not have a name");
+    }
+    return names_.front();
+  }
+
+  /**
+   * Return the first assigned name of the signal.
+   *
+   * \return First name of the signal
+   */
+  std::string name_or(std::string def) const {
+    if (!names_.size()) {
+      return def;
+    }
+    return names_.front();
+  }
+
+  /**
+   * Add a name of the signal.
+   *
+   * \param name Name of the signal
+   */
+  void add_name(std::string_view name) { names_.emplace_back(name); }
+
+ private:
+  /**
+   * Factory for Signal.
+   *
+   * \tparam T Type of the signal
+   * \param name Name of the signal
+   * \return owning unique pointer to Signal
+   * \note: Design-Note #1 reasons the existance of this factory
+   */
+  template <typename T>
+  static signal_ptr_t create() {
```

Comment:
This should return `std::unique_ptr<Signal>` as discussed in a previous comment.

Reply:
Although I'm also open to std::shared_ptr<Signal> if we know that it's always going to be used that way.

Reply:
Is there a reason to keep the connection alive by the data broker by explicitly storing the shared_ptr in the connections map? Otherwise you could (eg) also use weak_ptr and let the user decide when/if connections expire

Reply:
I think we'll discover this as we use it what works best.